sway/input/keyboard.c

```diff
@@ -9,88 +9,93 @@
 #include "sway/commands.h"
 #include "log.h"
 
-static bool keysym_is_modifier(xkb_keysym_t keysym) {
-	switch (keysym) {
-	case XKB_KEY_Shift_L: case XKB_KEY_Shift_R:
-	case XKB_KEY_Control_L: case XKB_KEY_Control_R:
-	case XKB_KEY_Caps_Lock:
-	case XKB_KEY_Shift_Lock:
-	case XKB_KEY_Meta_L: case XKB_KEY_Meta_R:
-	case XKB_KEY_Alt_L: case XKB_KEY_Alt_R:
-	case XKB_KEY_Super_L: case XKB_KEY_Super_R:
-	case XKB_KEY_Hyper_L: case XKB_KEY_Hyper_R:
-		return true;
-	default:
-		return false;
-	}
-}
-
-static size_t pressed_keysyms_length(xkb_keysym_t *pressed_keysyms) {
-	size_t n = 0;
-	for (size_t i = 0; i < SWAY_KEYBOARD_PRESSED_KEYSYMS_CAP; ++i) {
-		if (pressed_keysyms[i] != XKB_KEY_NoSymbol) {
-			++n;
+/**
+ * Update the shortcut model state in response to new input
+ */
+static void update_shortcut_model(struct sway_shortcut_state* state,
+				  struct wlr_event_keyboard_key * event,
+				  uint32_t new_key,
+				  bool last_key_was_a_modifier) {
```

Comment:
Note that some keycodes can yield one modifier keysym *and* on enon-modifier keysym.

Reply:
Technically there is no such thing as a "modifier keysym"; modifiers are associated to the current `xkb_state`, and the set of produced keysyms are associated to the current `xkb_state`. (Granted, the rule files governing the keymap sometimes link keysyms to modifiers; but in some cases (altwin) it links keycodes directly to modifiers. These linkages may also depend on other properties of the `xkb_state`; it's telling that the xkbcommon API never provides such an association. The mapping implied by `xmodmap` is, like its assumption that a keycode can produce no more than one keysym, incorrect in the general case.)